Content.Shared/_RMC14/Atmos/SharedRMCFlammableSystem.cs

```diff
@@ -339,6 +342,171 @@ private void OnFlammableIgnite(Entity<FlammableComponent> ent, ref RMCIgniteEven
+                    _entitiesNeedingRecheck.Remove(contact);
+                }
+            }
+        }
+        _fireContacts.Remove(ent.Owner);
+    }
+
+    private void OnExtinguishStartCollide(Entity<ExtinguishFireComponent> ent, ref StartCollideEvent args)
+    {
+        if ((args.OurFixture.CollisionLayer & (int)ent.Comp.Collision) == 0)
+            return;
+
+        TrackExtinguishContact(ent.Owner, args.OtherEntity);
+        TryExtinguishEntity(ent.Owner, args.OtherEntity);
+    }
+
+    private void OnExtinguishEndCollide(Entity<ExtinguishFireComponent> ent, ref EndCollideEvent args)
+    {
+        UntrackExtinguishContact(ent.Owner, args.OtherEntity);
+    }
+
+    private void OnExtinguishRemoved(Entity<ExtinguishFireComponent> ent, ref ComponentRemove args)
+    {
+        _extinguishContacts.Remove(ent.Owner);
+    }
+
+    private void CheckForExistingOverlaps(Entity<RMCIgniteOnCollideComponent> ent)
+    {
+        if (TerminatingOrDeleted(ent.Owner))
+            return;
+
+        var intersecting = _physics.GetEntitiesIntersectingBody(ent.Owner, (int)ent.Comp.Collision);
+
+        foreach (var entity in intersecting)
+        {
+            TrackFireContact(ent.Owner, entity);
+            TryIgnite(ent, entity, false);
+            if (_flammableQuery.HasComp(entity))
+            {
+                EnsureComp<SteppingOnFireComponent>(entity);
+                _entitiesNeedingRecheck.Add(entity);
+            }
+        }
+    }
+
+    private void TrackFireContact(EntityUid fireSource, EntityUid target)
+    {
+        if (!_fireContacts.TryGetValue(fireSource, out var contacts))
+        {
+            contacts = new HashSet<EntityUid>();
+            _fireContacts[fireSource] = contacts;
+        }
+
+        contacts.Add(target);
+    }
+
+    private void UntrackFireContact(EntityUid fireSource, EntityUid target)
+    {
+        if (_fireContacts.TryGetValue(fireSource, out var contacts))
+        {
+            contacts.Remove(target);
+            if (contacts.Count == 0)
+            {
+                _fireContacts.Remove(fireSource);
+            }
+        }
+    }
+
+    private void TrackExtinguishContact(EntityUid extinguishSource, EntityUid target)
+    {
+        if (!_extinguishContacts.TryGetValue(extinguishSource, out var contacts))
+        {
+            contacts = new HashSet<EntityUid>();
+            _extinguishContacts[extinguishSource] = contacts;
+        }
+        contacts.Add(target);
+    }
+
+    private void UntrackExtinguishContact(EntityUid extinguishSource, EntityUid target)
+    {
+        if (_extinguishContacts.TryGetValue(extinguishSource, out var contacts))
+        {
+            contacts.Remove(target);
+            if (contacts.Count == 0)
+                _extinguishContacts.Remove(extinguishSource);
+        }
+    }
+
+    private bool IsEntityTouchingFire(EntityUid entity)
+    {
+        foreach (var contacts in _fireContacts.Values)
```

Comment:
This should also just be getting tracked on the entity, this would be horrible for performance checking every single contact